packages/front-end/components/SchemaBrowser/EditSqlModal.tsx

```diff
@@ -54,45 +57,72 @@ export default function EditSqlModal({
   const [cursorData, setCursorData] = useState<null | CursorData>(null);
   const [testingQuery, setTestingQuery] = useState(false);
 
-  const handleTestQuery = useCallback(async () => {
-    const sql = form.getValues("sql");
-    try {
-      // Just check for basic SQL syntax, not any required columns
-      // We can check for required columns after the results are returned
-      validateSQL(sql, []);
+  const runTestQuery = async (sql: string) => {
+    validateSQL(sql, []);
+    setTestQueryResults(null);
+    const res: TestQueryResults = await apiCall("/query/test", {
+      method: "POST",
+      body: JSON.stringify({
+        query: sql,
+        datasourceId: datasourceId,
+      }),
+    });
+
+    if (res.results?.length && validateResponse) {
+      validateResponse(res.results[0]);
+    }
 
-      setTestingQuery(true);
-      const res = await apiCall<TestQueryResults>("/query/test", {
-        method: "POST",
-        body: JSON.stringify({
-          query: sql,
-          datasourceId,
-        }),
-      });
+    return res;
+  };
 
+  const handleTestQuery = async () => {
+    setTestingQuery(true);
+    try {
+      const sql = form.getValues("sql");
+      const res = await runTestQuery(sql);
       setTestQueryResults(res);
     } catch (e) {
       setTestQueryResults({ error: e.message });
     }
     setTestingQuery(false);
-  }, [form, apiCall, datasourceId]);
-
+  };
   const datasource = getDatasourceById(datasourceId);
-  // @ts-expect-error TS(2531) If you come across this, please fix it!: Object is possibly 'null'.
-  const supportsSchemaBrowser = datasource.properties.supportsInformationSchema;
+  const supportsSchemaBrowser =
+    datasource?.properties?.supportsInformationSchema;
 
   return (
     <Modal
       open
       header="Edit SQL"
       submit={form.handleSubmit(async (value) => {
+        if (testQueryBeforeSaving) {
+          // Reset test query results to avoid serving outdated errors/results
+          setTestQueryResults(null);
+          const res = await runTestQuery(value.sql);
+          if (res.error) {
+            throw new Error(res.error);
+          }
+        }
+
         await save(value.sql);
       })}
       close={close}
+      error={testQueryResults?.error}
```

Comment:
This is causing a bug.

To reproduce, create a new exp assign query and open the EditSqlModal - write valid sql and test it to make sure it works via the "Test Query". Remove one of the optional columns that isn't a dimension and click "Save". An error will be displayed. Now, add the missing column back and click "Test Query". The query will work, but the error message persists. It doesn't block the user from saving, but I can't figure out how to dismiss the error message.

Reply:
This is a weird one. It's happening because of two reasons:
- `runTestQuery` is throwing an error inside the `submit` fn ("\<column name\> doesn't exist"), which `Modal` is catching (and not `EditSqlModal`)
- `testQueryResults.error` never changes from `undefined` during this, so it never triggers a re-render

Check out the code in `Modal`: https://github.com/growthbook/growthbook/blob/mk/exp-assign-query/packages/front-end/components/Modal.tsx#L244-L256

The error caught from `submit` is overriding whatever error message is passed in from `EditSqlModal`. Since `EditSqlModal` technically never passes down anything new, `Modal` keeps its error state intact

Possible solution would be to catch any errors from `runTestQuery` and keep error message handling within `EditSqlModal`

Reply:
@bttf I can't thank you enough for taking a look at this and offering up some guidance.

Unfortunately, from what I can tell, if we want the form to not automatically close on submit in the event of an error, we have to throw the errors so they're ultimately caught by the `Modal`. If we don't pass them to the `Modal` it will just close on submit, even if there is an error, so I don't think we can keep the error handling within `EditSqlModal`.

With your help, I was able to come up with a solution to this bug, and I pushed it up in this [commit](https://github.com/growthbook/growthbook/pull/1370/commits/2a58de3c54f44da4b4a3d8cc700daad0f5661604), but.... I hate the implementation. 

It requires that I move the `testQueryResults` and `setTestQueryResults` out of `EditSqlModal` and into the parent component anywhere we render `EditSqlModal` AND right before I throw any of the errors, I call `setTestQueryResults` with the error - this way, whenever we re-test the sql (either via the "Test Query" or "Save" buttons) and we call `setTestResults(null)` the error is getting reset.

Very, very messy, and a terrible developer experience if anyone ever needs to do something similar to this in another area of the app.

I'm going to spend a few minutes looking at a different way to architect this that's cleaner - if you have any ideas, I'm all ears.

Reply:
I see what you mean ...  An alternative approach that is similarly hacky but much less verbose - Going back to the commit previous to your most recent one, I believe if you simply replaced `setQueryTestResults(res)` with 
```
setQueryTestResults({
  ...res,
  error: null
})
```
it would effectively reset the error message upon testing a query

Reply:
Interesting - that is definitely a much more concise change, but more than that, doesn't require hoisting the `testQueryResults` and `setTestQueryResults`.

Tested and it works as intended. Thank you, @bttf!